Dropout issues

#### Issue Description

I revisited issue #9801 and found that, while the dropout op can now be executed, there are four issues with it:

1. A `probabilityValue` of 0.0 should be permitted but an assertion fails, causing an exception.
```
INDArray in = Nd4j.ones(4, 8);
INDArray res = Nd4j.nn.dropout(in, false, 0.0); // throws exception
System.out.println(res);
```
2. Setting `inverted` to `true` should effectively change the dropout rate to `1 - probabilityValue` but nothing happens.
```
INDArray res1 = Nd4j.nn.dropout(in, false, 1.0);
System.out.println(res1); // same as res0 but should be different
INDArray res0 = Nd4j.nn.dropout(in, true, 1.0);
System.out.println(res0); // same as res1 but should be different
```
3. Input values that are _not_ dropped should remain unchanged but currently these values are divided by the `probabilityValue`.
```
INDArray res02 = Nd4j.nn.dropout(in, true, 0.2);
System.out.println(res02); // should only contain 1s and 0s but contains 5s instead of 1s
INDArray res08 = Nd4j.nn.dropout(in, true, 0.8);
System.out.println(res08); // should only contain 1s and 0s but contains 1.25s instead of 1s
```
4. Backpropagation through the dropout op causes an exception.
```
int batchSize = 4;
int seqLength = 8;

SameDiff sd = SameDiff.create();

SDVariable features = sd.placeHolder("features", DataType.FLOAT, batchSize, seqLength);
SDVariable labels = sd.placeHolder("labels", DataType.FLOAT, batchSize, batchSize);
SDVariable predictions = sd.nn.dropout("predictions", features, false, 0.5);
sd.loss.meanSquaredError("loss", labels, predictions, null);

TrainingConfig config = new TrainingConfig.Builder()
        .updater(new Adam(0.1))
        .dataSetFeatureMapping("features")
        .dataSetLabelMapping("labels")
        .build();
sd.setTrainingConfig(config);

RecordReader reader = new CollectionRecordReader(
        Collections.nCopies(batchSize, Collections.nCopies(seqLength + batchSize, new IntWritable(1))));
DataSetIterator iterator = new RecordReaderDataSetIterator(
        reader, batchSize, seqLength, seqLength + batchSize - 1, true);

System.out.println(sd.output(iterator, "predictions").get("predictions")); // forward pass works

sd.fit(iterator, 1); // backward pass throws exception
```

#### Version Information

* Deeplearning4j version: 1.0.0-SNAPSHOT
* Platform information (OS, etc): Linux Mint 20.3
* CUDA version, if used: N/A
* NVIDIA driver version, if in use: N/A

#### Additional Information

* Exception stack trace for bug 1:
```
Error at [/home/runner/work/deeplearning4j/deeplearning4j/libnd4j/include/ops/declarable/generic/random/dropout.cpp:44:0]:
dropout: Probability should be with range 0 to 1.
[main] ERROR org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner - Failed to execute op dropout. Attempted to execute with 1 inputs, 1 outputs, 1 targs,1 bargs and 1 iargs. Inputs: [(FLOAT,[4,8],c)]. Outputs: [(FLOAT,[4,8],c)]. tArgs: [0.0]. iArgs: [0]. bArgs: [false]. Op own name: "29b4b7a3-2537-4a14-ad03-eb797f300457" - Please see above message (printed out from c++) for a possible cause of error.
Exception in thread "main" java.lang.RuntimeException: Op [dropout] execution failed
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1561)
	at org.nd4j.linalg.factory.Nd4j.exec(Nd4j.java:6575)
	at org.nd4j.linalg.factory.ops.NDNN.dropout(NDNN.java:153)
	at Dropout.main(Dropout.java:23)
Caused by: java.lang.RuntimeException: Op with name 29b4b7a3-2537-4a14-ad03-eb797f300457 and op type [dropout] execution failed with message Op validation failed
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1905)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1540)
	... 4 more
```
* Exception stack trace for bug 4:
```
Exception in thread "main" java.lang.ClassCastException: class org.nd4j.linalg.api.ops.random.impl.DropOut cannot be cast to class org.nd4j.linalg.api.ops.CustomOp (org.nd4j.linalg.api.ops.random.impl.DropOut and org.nd4j.linalg.api.ops.CustomOp are in unnamed module of loader 'app')
	at org.nd4j.autodiff.samediff.serde.FlatBuffersMapper.fromFlatNode(FlatBuffersMapper.java:464)
	at org.nd4j.autodiff.samediff.serde.FlatBuffersMapper.cloneViaSerialize(FlatBuffersMapper.java:1065)
	at org.nd4j.autodiff.samediff.SameDiff.invokeGraphOn(SameDiff.java:656)
	at org.nd4j.autodiff.samediff.SameDiff.lambda$createGradFunction$3(SameDiff.java:4829)
	at org.nd4j.autodiff.samediff.SameDiff.defineFunction(SameDiff.java:4616)
	at org.nd4j.autodiff.samediff.SameDiff.defineFunction(SameDiff.java:4601)
	at org.nd4j.autodiff.samediff.SameDiff.createGradFunction(SameDiff.java:4821)
	at org.nd4j.autodiff.samediff.SameDiff.createGradFunction(SameDiff.java:4728)
	at org.nd4j.autodiff.samediff.SameDiff.fitHelper(SameDiff.java:1886)
	at org.nd4j.autodiff.samediff.SameDiff.fit(SameDiff.java:1808)
	at org.nd4j.autodiff.samediff.SameDiff.fit(SameDiff.java:1748)
	at org.nd4j.autodiff.samediff.config.FitConfig.exec(FitConfig.java:172)
	at org.nd4j.autodiff.samediff.SameDiff.fit(SameDiff.java:1728)
	at Dropout.main(Dropout.java:29)
```

--- a/libnd4j/include/ops/declarable/generic/random/dropout.cpp
+++ b/libnd4j/include/ops/declarable/generic/random/dropout.cpp
@@ -45,9 +45,9 @@ CONFIGURABLE_OP_IMPL(dropout, 1, 1, true, 1, 1) {
 
   if (block.width() > 1) reduceShape = INPUT_VARIABLE(1);
 
-  REQUIRE_TRUE(probValue > 0.f && probValue <= 1.f, 0, "dropout: Probability should be with range 0 to 1.");
+  REQUIRE_TRUE(probValue >= 0.f && probValue <= 1.f, 0, "dropout: Probability should be with range 0 to 1.");
 
-  if (probValue == 1.0) {
+  if (probValue == 1.0f) {
     *output = *input;
     return sd::Status::OK;
   }

--- a/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/serde/FlatBuffersMapper.java
+++ b/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/serde/FlatBuffersMapper.java
@@ -460,12 +460,14 @@ public class FlatBuffersMapper {
             //Set input SDVariables:
 
             //Set args:
-            //op.addTArgument();
-            ((CustomOp) op).addIArgument(extraInteger);
-            ((CustomOp) op).addTArgument(extraParams);
-            ((CustomOp) op).addBArgument(extraBools);
-            ((CustomOp) op).addDArgument(extraDTypes);
-            ((CustomOp) op).addSArgument(extraStrings);
+            if(op instanceof CustomOp) {
+                ((CustomOp) op).addIArgument(extraInteger);
+                ((CustomOp) op).addTArgument(extraParams);
+                ((CustomOp) op).addBArgument(extraBools);
+                ((CustomOp) op).addDArgument(extraDTypes);
+                ((CustomOp) op).addSArgument(extraStrings);
+            }
+
             //base loss gets saved as an int argument, ensure that the field is set
             if(op instanceof BaseLoss && extraInteger != null && extraInteger.length > 0) {
                 BaseLoss baseLoss = (BaseLoss) op;
@@ -476,8 +478,8 @@ public class FlatBuffersMapper {
             }
 
             op.setPropertiesForFunction(props);
-
-            ((CustomOp) op).configureFromArguments();
+            if(op instanceof CustomOp)
+                ((CustomOp) op).configureFromArguments();
             return op;
         } else {
             Class<?> c = LegacyOpMapper.getLegacyOpClassForId(opType, (int) opNum);

--- a/platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/rng/RandomTests.java
+++ b/platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/rng/RandomTests.java
@@ -23,10 +23,17 @@ package org.eclipse.deeplearning4j.nd4j.linalg.rng;
 
 import org.apache.commons.math3.random.JDKRandomGenerator;
 import org.apache.commons.math3.util.FastMath;
+import org.datavec.api.records.reader.RecordReader;
+import org.datavec.api.records.reader.impl.collection.CollectionRecordReader;
+import org.datavec.api.writable.IntWritable;
+import org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator;
 import org.junit.jupiter.api.*;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.MethodSource;
 
+import org.nd4j.autodiff.samediff.SDVariable;
+import org.nd4j.autodiff.samediff.SameDiff;
+import org.nd4j.autodiff.samediff.TrainingConfig;
 import org.nd4j.common.tests.tags.NativeTag;
 import org.nd4j.common.tests.tags.TagNames;
 import org.nd4j.linalg.BaseNd4jTestWithBackends;
@@ -55,15 +62,18 @@ import org.nd4j.linalg.api.rng.Random;
 import org.nd4j.linalg.api.rng.distribution.Distribution;
 import org.nd4j.linalg.api.rng.distribution.impl.NormalDistribution;
 import org.nd4j.linalg.api.rng.distribution.impl.OrthogonalDistribution;
+import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
 import org.nd4j.linalg.factory.Nd4j;
 import org.nd4j.linalg.factory.Nd4jBackend;
 import org.nd4j.linalg.indexing.BooleanIndexing;
 import org.nd4j.linalg.indexing.conditions.Conditions;
+import org.nd4j.linalg.learning.config.Adam;
 import org.nd4j.rng.NativeRandom;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -229,6 +239,62 @@ public class RandomTests extends BaseNd4jTestWithBackends {
 
         assertEquals(z1, z2);
     }
+
+    @ParameterizedTest
+    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
+    public void testDropoutZero(Nd4jBackend backend) {
+        INDArray in = Nd4j.ones(4, 8);
+        INDArray res = Nd4j.nn.dropout(in, false, 0.0); // throws exception
+        assertEquals(0.0,res.sumNumber().doubleValue(),1e-6);
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
+    public void testDropoutOne(Nd4jBackend backend) {
+        INDArray in = Nd4j.ones(4, 8);
+        INDArray res1 = Nd4j.nn.dropout(in, false, 1.0);
+        System.out.println(res1); // same as res0 but should be different
+        INDArray res0 = Nd4j.nn.dropout(in, true, 1.0);
+        System.out.println(res0); // same as res1 but should be different
+        assertFalse(res1.eq(res0).all());
+
+    }
+
+
+    @ParameterizedTest
+    @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")
+    public void testDropoutBackPropRnn(Nd4jBackend backend) {
+        int batchSize = 4;
+        int seqLength = 8;
+
+        SameDiff sd = SameDiff.create();
+
+        SDVariable features = sd.placeHolder("features", DataType.FLOAT, batchSize, seqLength);
+        SDVariable labels = sd.placeHolder("labels", DataType.FLOAT, batchSize, batchSize);
+        SDVariable random = sd.var(Nd4j.random().uniform(0.0,1.0,DataType.FLOAT,batchSize, batchSize));
+        SDVariable predictions = sd.nn.dropout("predictions", features, false, 0.5);
+        sd.loss.meanSquaredError("loss", labels, random, null);
+
+        TrainingConfig config = new TrainingConfig.Builder()
+                .updater(new Adam(0.1))
+                .dataSetFeatureMapping("features")
+                .dataSetLabelMapping("labels")
+                .build();
+        sd.setTrainingConfig(config);
+
+        RecordReader reader = new CollectionRecordReader(
+                Collections.nCopies(batchSize, Collections.nCopies(seqLength + batchSize, new IntWritable(1))));
+        DataSetIterator iterator = new RecordReaderDataSetIterator(
+                reader, batchSize, seqLength, seqLength + batchSize - 1, true);
+
+        System.out.println(sd.output(iterator, "predictions").get("predictions")); // forward pass works
+        //ensure backprop also works
+        sd.fit(iterator, 1); // backward pass throws exception
+
+    }
+
+
 
     @ParameterizedTest
     @MethodSource("org.nd4j.linalg.BaseNd4jTestWithBackends#configs")

